### PR TITLE
Feature: apriltag diagnostics

### PR DESCRIPTION
**Pull request by jrepp**
This change adds compile-time diagnostic and experiments support and runtime traceability of each classification stage to the apriltag library.

Motivation
* Build understanding of the internals of the library
* Support real-time capture of internal detection phases such as cluster finding and quad building
* Support posthoc analysis of math routines for verification and optimization
* Support configuration options at build time to enable experiments and optimization
* Keep the library internals intact for parity with upstream

Description of Changes
* Lift & shift of the debug output routines: all if(debug) sections that emit files have been moved to the apriltag_demo program. The library has an individual callback for each distinct debug stage. I have verified file-for-file and bit-for-bit output (except in the case of floating-point to text ambiguity in the .ps output) 
* Move all non-library essential bits into the example/ folder
* Set of debug, diagnostic and developer support flags added to CMakeLists.txt
* apriltag_config.h.in -> apriltag_config.h file added using standard CMake convention
* Make time profiling, asserts and diagnostic hooks optional at compile time
* It is no longer allowed to enable 'debug' from python - a further change should add the diagnostic structure callback to the python library for research support

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,6 @@
 example/opencv_demo
 example/apriltag_demo
 build/
-
+build-debug/
+build-release/
+.idea/
```
Comment (christian-rauch) [inline .gitignore:10]: Those are IDE specific. I am in favour of not adding all IDE-specific files here, but only keeping those that are produced by the documented make or make procedure.

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,20 +32,63 @@ endif()
...
+#
+option(AT_DEBUG_GAUSSIAN "Debug the gaussian blur kernel" OFF)
+
+#
+# Developer test flags (experimental features)
+#
+option(AT_TEST_THRESHOLD_DEGLITCH "Enable threshold deglitching for noisy images" OFF)
+
+option(AT_HOMOGRAPHY_METHOD_INVERSE "Use inverse method" ON)
+cmake_dependent_option(AT_HOMOGRAPHY_METHOD_CHO_OR_PLU "Use CHO matrix solver" OFF AT_HOMOGRAPHY_METHOD_INVERSE CHO)
```
Comment (christian-rauch) [inline CMakeLists.txt:56]: Can you move those up to the other build options? Let's keep all the options in one place, so one has not to search through all of the CMakeLists.txt.

**Review by christian-rauch — CHANGES_REQUESTED** [review]

**Comment by christian-rauch** [comment]
This is a huge pull request. I think you will have to split this up into multiple PRs. There are many diagnostics unrelated changes, such as pure code style and documentation fixes and a restructuring of the file hierarchy. At least those have to be split into a dedicated PR. It is unlikely anyone will have the time to review this PR if there are soo many distinct changes touching nearly every file in the repo.

File: .gitignore (hunk 1 above)
Comment (jrepp) [reply .gitignore]: the flavor builds are not IDE specific, are you ok those being simplified to build-*/?
I removed .idea - also I just learned about `git config --global core.excludesfile ~/.gitignore_global`

File: CMakeLists.txt (hunk 2 above)
Comment (jrepp) [reply CMakeLists.txt:56]: How do you feel if I group options together and then the triggering of those options (e.g. ASAN, if(ASAN)) separately? We can either have the option and its trigger together or all options then all triggers.

Comment (christian-rauch) [reply CMakeLists.txt:56]: I would prefer to have the options grouped together. So when someone opens the `CMakeLists.txt` to look for the options, they only have to scroll to one section.

**Comment by jrepp** [comment]
> This is a huge pull request. I think you will have to split this up into multiple PRs. There are many diagnostics unrelated changes, such as pure code style and documentation fixes and a restructuring of the file hierarchy. At least those have to be split into a dedicated PR. It is unlikely anyone will have the time to review this PR if there are soo many distinct changes touching nearly every file in the repo.

I agree it touches a lot of files and there are file moves, it's mostly down to the nature of specific change underway here. I can go back to the drawing board to see if I can introduce this in smaller PRs but I can see that is going to be difficult to pull off. Perhaps I can undo the file moves.

I should have pointed out as well that this is a very mechanical transformation that can be A/B tested (see below).

The most difficult code to evaluate for correctness is the lifting and shifting of `if(td->debug)` sections. In order to validate my changes, I have the following test procedure.

For core changes
Run opencv_demo with address sanitation and all diagnostics - I use the [36h11 calibration image](https://drive.google.com/file/d/18hqYiBIBrhel1-ybMzMrOU9vWIfsutTR/view?usp=sharing) on a printed piece of paper and the CLion profiler to check for performance regressions.

Run apriltag_demo with the [test image nori-5-640.jpg](https://drive.google.com/file/d/1E314SxrRXUh47GUYbEpQvqdLBELXNNtn/view?usp=sharing)  - this is my cat Nori with the tag value 5 represented in 36h11`

```
loading /home/jrepp/Downloads/nori-5-640.jpg
[W] [../example/pjpeg.c:674] Unknown marker ed at offset 7edd
image: /home/jrepp/Downloads/nori-5-640.jpg 480x640
detection   0: id (36x11)-5   , hamming 0, margin  124.404
 0                             init        0.000000 ms        0.000000 ms
 1                         decimate        0.330000 ms        0.330000 ms
 2                       blur/sharp        0.152000 ms        0.482000 ms
 3                        threshold        2.190000 ms        2.672000 ms
 4                        unionfind        6.867000 ms        9.539000 ms
 5                    make clusters        5.106000 ms       14.645000 ms
 6            fit quads to clusters       12.684000 ms       27.329000 ms
 7                            quads        2.145000 ms       29.474000 ms
 8                decode+refinement        3.073000 ms       32.547000 ms
 9                        reconcile        0.006000 ms       32.553000 ms
10                     debug output       44.032000 ms       76.585000 ms
11                          cleanup        0.004000 ms       76.589000 ms
hamm     1     0     0     0     0     0     0     0     0     0 time       76.589ms quads    18
==== All Iterations:
hamm     1     0     0     0     0     0     0     0     0     0 time       76.589ms quads    18
```

Then I run the same program using the clean unmodified repository and md5 all of the outputs:

running debug build against upstream/master
```
ls -1 debug_*|xargs -n 1 md5sum
7be3f4b9a5b0f89147dea510072bd385  debug_clusters.pnm
38ce36fab4190564a386734939fdb4e3  debug_lines.ps
4bd611a32f109a53158d244906ec2fbd  debug_output.pnm
17ab40449db304a00647e302ef97ae49  debug_output.ps
fca4a809c7c3b53c2c41a42b929a1560  debug_preprocess.pnm
64b11535fe10257c101a08f5524a539e  debug_quads_fixed.pnm
29b65193bbf1eaad4f5b55c3d02a31af  debug_quads.ps
706c1aa48ed04f7a9135dbef812ac2e9  debug_quads_raw.pnm
5dfbdd1bc284cc62141d6d9c2e91ae27  debug_samples.pnm
2108e5ed5cb4d819734340384913284d  debug_segmentation.pnm
fef723f18f652dd14f6fdabbb6902277  debug_threshold.pnm
```

Comparing the results:
```
diff /tmp/debug.md5sum.txt /tmp/debug.md5sum-2.txt 
3,4c3,4
< 4bd611a32f109a53158d244906ec2fbd  debug_output.pnm
< 17ab40449db304a00647e302ef97ae49  debug_output.ps
---
> 12fb20e06b7d7a4243fa70dc84cdeea7  debug_output.pnm
> 500f1dd3a175ed07a2bf9be5a1a54ae9  debug_output.ps
7c7
< 29b65193bbf1eaad4f5b55c3d02a31af  debug_quads.ps
---
> 1729e24a924cf3d4bc5a2d01e9b8300c  debug_quads.ps
```
The visual differences are due to the random color selection and some float to text conversion ambiguity.

Here's the upstream vs diagnostic version of quads.ps
image

**Comment by christian-rauch** [comment]
> I agree it touches a lot of files and there are file moves, it's mostly down to the nature of specific change underway here. I can go back to the drawing board to see if I can introduce this in smaller PRs but I can see that is going to be difficult to pull off. Perhaps I can undo the file moves.

If you have moved those changes into separate commits, then splitting them out into another PR should be easy. You also have quite a few commits. Some of them appear to be "fixup" commits that fix issues from your earlier commits. Those should be squashed so that there is a single commit that implements a function correctly without additional fixes later. This all is important so that the history changes is properly documented.

> The most difficult code to evaluate for correctness is the lifting and shifting of if(td->debug) sections. In order to validate my changes, I have the following test procedure.

While these procedures are good to make sure that the original approach does not break, the PRs must be clearly separated in order to be easier to "digest". With so many changes, it is very difficult to trace the changes that only change the style, fix warnings etc, and those that actually add a feature or change the behaviour.

I strongly suggest that you submit the style/warning/etc changes, that do not change the actual implementation, first and get them reviewed and merged. After this, you can rebase your "feature PR" (i.e. this very PR) on top of the new upstream branch so that we can review only the diagnostics changes separately.

**Comment by jrepp** [comment]
I have posted small cleanup change lists and an initial diagnostic option patch for assert. I would like to have the apriltag_config.h support for additional diagnostic / developer options. https://github.com/AprilRobotics/apriltag/pull/211

**Pull request closed without merging**